cpp/include/cudf/io/arrow_io_source.hpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "datasource.hpp"
+
+#include <arrow/buffer.h>
+
+// We disable warning 611 because some Arrow subclasses of
+// `arrow::fs::FileSystem` only partially override the `Equals` method,
+// triggering warning 611-D from nvcc.
+#ifdef __CUDACC__
+#pragma nv_diag_suppress 611
+#endif
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/s3fs.h>
+#ifdef __CUDACC__
+#pragma nv_diag_default 611
+#endif
+
+// We disable warning 2810 to workaround the compile issue (warning treated as error):
+// result.h(263): error #2810-D: ignoring return value type with "nodiscard" attribute
+#ifdef __CUDACC__
+#pragma nv_diag_suppress 2810
+#endif
+#include <arrow/result.h>
+#ifdef __CUDACC__
+#pragma nv_diag_default 2810
+#endif
```

Comment:
We've recently upgraded to arrow 12. Can you try removing these suppression please?

Reply:
Testing locally your latest changes in https://github.com/rapidsai/cudf/pull/13698/commits/031ca64f5fab857befc650ed61e0d12aa2d1bf73.

Reply:
The test build was successful. So we may be safe to remove such suppressions.

Reply:
Let's see what CI says :)